### DISCUSSION TITLE
Calibrating Touchscreens

### DISCUSSION THREAD

**Discussion by Jean-Gabriel1982** [GrandOrgue/grandorgue#736] [General]
Hello all,
I’m finishing my GO configuration and have some troubles with my touchscreens. 

I have two Elo touchscreens right and left my keyboards. But calibration is not effective mouse pointer is about 10cm right or left from my finger. 

I une Linux Mint 20.2 with kernel 5.11. 

I use some commands like :
- xinput
And 
- xrandr
And 
- xinput map-to-output <device> <output>

after that calibration is ok and I can use my two screens correctly, but when I restart my computer, all those parameters are lost. 

Do you know how to keep the parameters persistent ?

Many thanks for your help :-)

**Comment by larspalo** [comment]
This sounds like some xorg.conf stuff to modify... Have you looked at xinput_calibrator?

**Reply by Jean-Gabriel1982** [reply]
> This sounds like some xorg.conf stuff to modify... Have you looked at xinput_calibrator?

Yes I tried with Xinput calibrator, but the program considers the two screens as one, and therefore calibration is impossible. I also tried to install the screen drivers, but this is the same, the calibration does not take to restart the computer ...

**Comment by oleg68** [comment]
I also have two same touchscreens. For right calibration I call the script from the GO staring script

```
#!/usr/bin/bash

DEVICE='Silicon Integrated System Co. SiS HID Touch Controller'
LEFT_PATH=pci-0000:00:14.0-usb-0:3.3.3:1.0-event
RIGHT_PATH=pci-0000:00:14.0-usb-0:3.2.3:1.0-event
LEFT_MON=DVI-D-0
RIGHT_MON=DVI-D-1

path_2_event_path() {
	readlink -f /dev/input/by-path/$1
}

LEFT_EVENT=$(path_2_event_path $LEFT_PATH)
RIGHT_EVENT=$(path_2_event_path $RIGHT_PATH)

list_touchscreens() {
	xinput list | grep "$DEVICE" | sed -n 's/.*id=\([0-9]\+\).*/\1/p'
}

event_path() {
	INPUT_ID=$1
	xinput list-props $INPUT_ID | awk -F: '/Device Node/ { print $2;}' | xargs
}

for INPUT_ID in $(list_touchscreens); do
	EVENT=$(event_path $INPUT_ID)
	MON=
	[[ "$EVENT" == "$LEFT_EVENT" ]] && MON=$LEFT_MON
	[[ "$EVENT" == "$RIGHT_EVENT" ]] && MON=$RIGHT_MON
	xinput map-to-output $INPUT_ID $MON
done
```

**Reply by Jean-Gabriel1982** [reply]
🥺

**Reply by Jean-Gabriel1982** [reply]
> I also have two same touchscreens. For right calibration I call the script from the GO staring script
> 
> ```
> #!/usr/bin/bash
> 
> DEVICE='Silicon Integrated System Co. SiS HID Touch Controller'
> LEFT_PATH=pci-0000:00:14.0-usb-0:3.3.3:1.0-event
> RIGHT_PATH=pci-0000:00:14.0-usb-0:3.2.3:1.0-event
> LEFT_MON=DVI-D-0
> RIGHT_MON=DVI-D-1
> 
> path_2_event_path() {
> 	readlink -f /dev/input/by-path/$1
> }
> 
> LEFT_EVENT=$(path_2_event_path $LEFT_PATH)
> RIGHT_EVENT=$(path_2_event_path $RIGHT_PATH)
> 
> list_touchscreens() {
> 	xinput list | grep "$DEVICE" | sed -n 's/.*id=\([0-9]\+\).*/\1/p'
> }
> 
> event_path() {
> 	INPUT_ID=$1
> 	xinput list-props $INPUT_ID | awk -F: '/Device Node/ { print $2;}' | xargs
> }
> 
> for INPUT_ID in $(list_touchscreens); do
> 	EVENT=$(event_path $INPUT_ID)
> 	MON=
> 	[[ "$EVENT" == "$LEFT_EVENT" ]] && MON=$LEFT_MON
> 	[[ "$EVENT" == "$RIGHT_EVENT" ]] && MON=$RIGHT_MON
> 	xinput map-to-output $INPUT_ID $MON
> done
> ```

I solved my calibration issue. I just copy the 2 commands map to output in a txt file, and I put the txt file in program running at startup. 

The calibration is good now. 
but the problem I have is I can click on a menu but I can’t click on a submenu. For example, in GO I can see the main menu but I can’t open Settings, nothing happens when I put my finger on it. 

It is not GO ;-) it doesn’t work whatever program I use ..

**Reply by oleg68** [reply]
May be it is still a calibration issue?

Try to move the GO main window from the top of screen somewere in the middle. Will you be still able to click on a menu?

**Reply by Jean-Gabriel1982** [reply]
> May be it is still a calibration issue?
> 
> Try to move the GO main window from the top of screen somewere in the middle. Will you be still able to click on a menu?

Yes I tried, same everywhere whatever screen I use. 
i can clic on main menus and develop them, but clicking on submenus doesn’t not work

**Reply by Jean-Gabriel1982** [reply]
> 
> 
> May be it is still a calibration issue?
> 
> Try to move the GO main window from the top of screen somewere in the middle. Will you be still able to click on a menu?

Maybe it is a problem on my Linux Distribution ... i installed Linux Mint XFCE Editions 20.2 UMA.
Maybe i have to install another Distribution ? Or Linux mint Cinnamon ?
My brother advised me for Mint as it seems to be a light distribution (faster startup ..)

**Reply by oleg68** [reply]
I've just tested my touchscreens. I can click on both menus and submenus on Fedora/Gnome. Maybe it is a XFCE issue or a problem with your touchscreen driver?

**Reply by Jean-Gabriel1982** [reply]
I think it is xfce. I will install Cinnamon this evening.. hope this will work
https://linuxmint.com/edition.php?id=288

**Comment by Jean-Gabriel1982** [comment]
@oleg68 I does not work .. i hope my screens are good pffff :(

Maybe i have to change Linux .. could you advise me ? 
Solus ? Ubuntu ? Nitrux ? Elementary OS ? Tails ?
Are thos distribution can natively have drivers ? Maybe my screens are too old .. Elo 1723L ...

I'm so disapointed ... sorry for asking so many things :/

**Comment by oleg68** [comment]
> Maybe i have to change Linux .. could you advise me ?
Solus ? Ubuntu ? Nitrux ? Elementary OS ? Tails ?

At first, try to install Gnome on your Mint. If it still  doesn't work, than it is to try to configure.

> Are thos distribution can natively have drivers ? 

Malprobably. Usually drivers are the same in all distros.

**Reply by Jean-Gabriel1982** [reply]
> > Maybe i have to change Linux .. could you advise me ?
> > Solus ? Ubuntu ? Nitrux ? Elementary OS ? Tails ?
> 
> At first, try to install Gnome on your Mint. If it still doesn't work, than it is to try to configure.
> 
> > Are thos distribution can natively have drivers ?
> 
> Malprobably. Usually drivers are the same in all distros.

I’ve installed Ubuntu with Gnome 3
I had to put again in a txt file at startup the commands for calibrating the touchscreens. 

It works fine …. Except on GrandOrgue :D
It’s not a really problem. I can take mouse for menus (in don’t understand why it is only on GO) but I can select the stops fine. 

BUT ! …. :-) …. Another big problem. When I want to select a stop, a virtual keyboard appears on the screens. In Accesibility it is not selected. I can’t désactivate it ..
Have you an idea ?

**Reply by oleg68** [reply]
> When I want to select a stop, a virtual keyboard appears on the screens. In Accesibility it is not selected. I can’t désactivate it .. Have you an idea ?

There is a [gnome extension](https://github.com/keringar/cariboublocker) that disables this behaviour. Try to install it.

**Reply by Jean-Gabriel1982** [reply]
> There is a [gnome extension](https://github.com/keringar/cariboublocker) that disables this behaviour. Try to install it.

Ah i see that, but extracting i have two files ... i don't know what to do with that ..

**Comment by oleg68** [comment]
> I can take mouse for menus (in don’t understand why it is only on GO) but I can select the stops fine.

Could you clarify your problem? What doesn't work exactly?

**Reply by Jean-Gabriel1982** [reply]
> Could you clarify your problem? What doesn't work exactly?

Only on GO, when it is open, i can click on a menu, but not a submenu
For example File>Settings
I can click on File, but if is click on Settings, nothing happen.

**Reply by oleg68** [reply]
> Only on GO, when it is open, i can click on a menu, but not a submenu

With mouse, with touchscreen or with both?

Does the current behaviour differ from the original one on Mint/LXDE?

**Reply by Jean-Gabriel1982** [reply]
> 
> 
> > Only on GO, when it is open, i can click on a menu, but not a submenu
> 
> With mouse, with touchscreen or with both?

With touchscreen only. With mouse it is ok.

> Does the current behaviour differ from the original one on Mint/LXDE?

Yes, in Mint i couldn't open any submenu on the system (touchscreen only)

**Comment by oleg68** [comment]
@Jean-Gabriel1982 

> Ah i see that, but extracting i have two files ... i don't know what to do with that ..

Please take a look at https://github.com/GrandOrgue/grandorgue/discussions/615

**Reply by Jean-Gabriel1982** [reply]
> 
> 
> @Jean-Gabriel1982
> 
> > Ah i see that, but extracting i have two files ... i don't know what to do with that ..
> 
> Please take a look at #615

Thank you ! i'll check this evening :)

**Reply by Jean-Gabriel1982** [reply]
> @Jean-Gabriel1982
> 
> > Ah i see that, but extracting i have two files ... i don't know what to do with that ..
> 
> Please take a look at #615

I don't success ... I find how to copy the folder on the folder gnome-shell/extensions ... but step 5 ... i don't know how to enable ... it tells ""-extensions" command not found"

**Reply by oleg68** [reply]
> I don't success ... I find how to copy the folder on the folder gnome-shell/extensions ... but step 5 ... i don't know how to enable ... it tells ""-extensions" command not found"

1. Open a terminal window
2. Run the command

    ```
    gnome-extensions enable block-caribou-36@lxylxy123456.ercli.dev
    ```
3. Restart gnome. The best way is to log out and then to log in again

**Reply by Jean-Gabriel1982** [reply]
Yes I put the command yesterday but the terminal returns « -extensions » not found

**Reply by oleg68** [reply]
> Yes I put the command yesterday but the terminal returns « -extensions » not found

You didn't type `gnome` in the command

**Reply by Jean-Gabriel1982** [reply]
> > Yes I put the command yesterday but the terminal returns « -extensions » not found
> 
> You didn't type `gnome` in the command

Yes I’ve typed it … but my terminal was opened at the wrong place .. now it works !
Thank you for your patience :-)

**Comment by oleg68** [comment]
@Jean-Gabriel1982 

> > 3. Does running the calibration script manually solve the problems?
> 
> It seems not.
> I run on my terminal the two commands
> xinput map-to-output 9 HDMI-A-1
> xinput map-to-output 11 HDMI-A-0
> 
> 
> But there is no change

The reason is that the numbers 9 and 11 may be changed across the reboots, so they must not be hardcoded. The calibrating script have to calculate them.

See [my script above](https://github.com/GrandOrgue/grandorgue/discussions/736#discussioncomment-1397808) as an example.

**Reply by Jean-Gabriel1982** [reply]
@oleg68 
I read that the Touchscreen configuration has to be saved in a Xorg file or something .. i don't understand as you can imagine :D
Does your script replace that ?

So we have just to save your script in a text file and call this file at computer startup ?

**Comment by oleg68** [comment]
@Jean-Gabriel1982 

I read that the Touchscreen configuration has to be saved in a Xorg file or something .. i don't understand as you can imagine :D Does your script replace that ?

Yes, it does.

> So we have just to save your script in a text file and call this file at computer startup ?

1. This script has to be adapted to your environment: the five variables on top of the script
2. It is better to call this script from the script for starting GO

**Comment by Jean-Gabriel1982** [comment]
Indeed the ID of the screens has changed
>~$ xinput list
⎡ Virtual core pointer                    	id=2	[master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer              	id=4	[slave  pointer  (2)]
⎜   ↳ Logitech USB Optical Mouse              	id=9	[slave  pointer  (2)]
⎜   ↳ eGalax Inc. Elo Touch Solutions Pcap USB Interface UNKNOWN	id=13	[slave  pointer  (2)]
⎜   ↳ eGalax Inc. Elo Touch Solutions Pcap USB Interface UNKNOWN	id=14	[slave  pointer  (2)]
⎜   ↳ eGalax Inc. Elo Touch Solutions Pcap USB Interface	id=15	[slave  pointer  (2)]
⎜   ↳ eGalax Inc. Elo Touch Solutions Pcap USB Interface	id=16	[slave  pointer  (2)]
⎣ Virtual core keyboard                   	id=3	[master keyboard (2)]
    ↳ Virtual core XTEST keyboard             	id=5	[slave  keyboard (3)]
    ↳ Power Button                            	id=6	[slave  keyboard (3)]
    ↳ Video Bus                               	id=7	[slave  keyboard (3)]
    ↳ Power Button                            	id=8	[slave  keyboard (3)]
    ↳ Logitech USB Keyboard System Control    	id=12	[slave  keyboard (3)]
    ↳ Logitech USB Keyboard                   	id=10	[slave  keyboard (3)]
    ↳ Logitech USB Keyboard Consumer Control  	id=11	[slave  keyboard (3)]`

2/ You mean i have to copy all the script and add it to the "pasuspender /bin/GrandOrgue" in my starting program ??

**Comment by oleg68** [comment]
> Indeed the ID of the screens has changed

Obviously, 

```
DEVICE=' eGalax Inc. Elo Touch Solutions Pcap USB Interface'
```

But it also needs to specify proper usb paths in `LEFT_PATH` and `RIGHT_PATH`

For obtaining them at first get the event id:

```
xinput list-props _DisplaiId_ | grep 'Device Node'
```

and then find it among the result of

```
ls -l /dev/input/by-path/
```

At last `LEFT_MON` and `RIGHT_MON` should be set to your ids `HDMI-A-1` and  `HDMI-A-0`


> 
> 2/ You mean i have to copy all the script and add it to the "pasuspender /bin/GrandOrgue" in my starting program ??

I recommend you to create a text shell script for running GO. It should
1. Make a sleep 5
2. Calibrate the touchscreens
3. pasuspender /usr/bin/GrandOrgue

**Reply by Jean-Gabriel1982** [reply]
I'm sorry but i'm completely lost.
> unable to find _DisplaiID_

**Reply by oleg68** [reply]
Substitute your number instead of _DisplayId_

**Reply by Jean-Gabriel1982** [reply]
@oleg68 
Good morning, thank you for your help !
Here is the result from the terminal

> xinput list
⎡ Virtual core pointer                    	id=2	[master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer              	id=4	[slave  pointer  (2)]
⎜   ↳ eGalax Inc. Elo Touch Solutions Pcap USB Interface UNKNOWN	id=10	[slave  pointer  (2)]
⎜   ↳ Logitech USB Optical Mouse              	id=13	[slave  pointer  (2)]
⎜   ↳ eGalax Inc. Elo Touch Solutions Pcap USB Interface UNKNOWN	id=12	[slave  pointer  (2)]
⎜   ↳ eGalax Inc. Elo Touch Solutions Pcap USB Interface	id=9	[slave  pointer  (2)]
⎜   ↳ eGalax Inc. Elo Touch Solutions Pcap USB Interface	id=11	[slave  pointer  (2)]
⎣ Virtual core keyboard                   	id=3	[master keyboard (2)]
    ↳ Virtual core XTEST keyboard             	id=5	[slave  keyboard (3)]
    ↳ Power Button                            	id=6	[slave  keyboard (3)]
    ↳ Video Bus                               	id=7	[slave  keyboard (3)]
    ↳ Power Button                            	id=8	[slave  keyboard (3)]
    ↳ Logitech USB Keyboard                   	id=14	[slave  keyboard (3)]
    ↳ Logitech USB Keyboard Consumer Control  	id=15	[slave  keyboard (3)]
    ↳ Logitech USB Keyboard System Control    	id=16	[slave  keyboard (3)]
jean-gabriel@jeangabriel-B450-I-AORUS-PRO-WIFI:~$ xinput list-props _9_ | grep 'Device Node'
unable to find device _9_
jean-gabriel@jeangabriel-B450-I-AORUS-PRO-WIFI:~$ xinput list-props 9 | grep 'Device Node'
	Device Node (289):	"/dev/input/event4"
jean-gabriel@jeangabriel-B450-I-AORUS-PRO-WIFI:~$ ls -l /dev/input/by-path/
total 0
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:1.5:1.0-event -> ../event4
lrwxrwxrwx 1 root root 10 nov.  22 08:37 pci-0000:01:00.0-usb-0:1.5:1.0-event-mouse -> ../event20
lrwxrwxrwx 1 root root  6 nov.  22 08:37 pci-0000:01:00.0-usb-0:1.5:1.0-mouse -> ../js0
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:2.5:1.0-event -> ../event5
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:2.5:1.0-event-mouse -> ../event6
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:2.5:1.0-mouse -> ../mouse3
lrwxrwxrwx 1 root root 10 nov.  22 08:37 pci-0000:01:00.0-usb-0:3.1:1.0-event-mouse -> ../event10
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:3.1:1.0-mouse -> ../mouse4
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:4:1.0-event-kbd -> ../event7
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:4:1.1-event -> ../event8
lrwxrwxrwx 1 root root  9 nov.  22 08:37 platform-pcspkr-event-spkr -> ../event3
jean-gabriel@jeangabriel-B450-I-AORUS-PRO-WIFI:~$ xinput list-props 11 | grep 'Device Node'
	Device Node (289):	"/dev/input/event5"
jean-gabriel@jeangabriel-B450-I-AORUS-PRO-WIFI:~$ ls -l /dev/input/by-path/
total 0
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:1.5:1.0-event -> ../event4
lrwxrwxrwx 1 root root 10 nov.  22 08:37 pci-0000:01:00.0-usb-0:1.5:1.0-event-mouse -> ../event20
lrwxrwxrwx 1 root root  6 nov.  22 08:37 pci-0000:01:00.0-usb-0:1.5:1.0-mouse -> ../js0
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:2.5:1.0-event -> ../event5
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:2.5:1.0-event-mouse -> ../event6
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:2.5:1.0-mouse -> ../mouse3
lrwxrwxrwx 1 root root 10 nov.  22 08:37 pci-0000:01:00.0-usb-0:3.1:1.0-event-mouse -> ../event10
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:3.1:1.0-mouse -> ../mouse4
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:4:1.0-event-kbd -> ../event7
lrwxrwxrwx 1 root root  9 nov.  22 08:37 pci-0000:01:00.0-usb-0:4:1.1-event -> ../event8
lrwxrwxrwx 1 root root  9 nov.  22 08:37 platform-pcspkr-event-spkr -> ../event3

**Reply by Jean-Gabriel1982** [reply]
With your screen, i modify for my parameters .. i'll check :)

> #!/usr/bin/bash

DEVICE='eGalax Inc. Elo Touch Solutions Pcap USB Interface'
LEFT_PATH=pci-0000:01:00.0-usb-0:1.5:1.0-event
RIGHT_PATH=pci-0000:01:00.0-usb-0:2.5:1.0-event
LEFT_MON=HDMI-A-1
RIGHT_MON=HDMI-A-0

path_2_event_path() {
	readlink -f /dev/input/by-path/$1
}

LEFT_EVENT=$(path_2_event_path $LEFT_PATH)
RIGHT_EVENT=$(path_2_event_path $RIGHT_PATH)

list_touchscreens() {
	xinput list | grep "$DEVICE" | sed -n 's/.*id=\([0-9]\+\).*/\1/p'
}

event_path() {
	INPUT_ID=$1
	xinput list-props $INPUT_ID | awk -F: '/Device Node/ { print $2;}' | xargs
}

for INPUT_ID in $(list_touchscreens); do
	EVENT=$(event_path $INPUT_ID)
	MON=
	[[ "$EVENT" == "$LEFT_EVENT" ]] && MON=$LEFT_MON
	[[ "$EVENT" == "$RIGHT_EVENT" ]] && MON=$RIGHT_MON
	xinput map-to-output $INPUT_ID $MON
done

**Comment by emseedee** [comment]
Have you tried using the ELO Linux touch screen drivers, from https://www.elotouch.co.uk/support/downloads#/category/

I'm using 2 ELOtouch ET1715L monitors under Ubuntu 18.04 and they work fine, although I do sometimes need to recalibrate after a software update, and also sometimes need to restart the PC after an update, so that the system picks up the calibration information.  

By the way - the reason that I'm using an earlier version of Ubuntu is that the ELO drivers don't seem to support the later versions yet.

**Reply by Jean-Gabriel1982** [reply]
Of course I've downloaded their driver, and do what they explain step to step. 

At the end, absolutely nothing changes. And they don't answer to emails ..

After that as I am a very very beginner in Linux, maybe I made a mistake somewhere.. but I read carefully their "installation instructions" and follow step by step ..

**Reply by emseedee** [reply]
Yes, they're not very good at replying to emails - they did a couple of years ago but now they've stopped. 

I found that running the elova script works in Ubuntu 18.06, but not in later versions.  When I tried to run it on Ubuntu 20, it didn't do anything - so perhaps this is what you're seeing.

I've just looked at the latest version of the drivers, and it's all a bit odd - I'm running version 5.1.0, dated Mar 25, 2019, and the current version seems to be 4.1.0, but dated May 4, 2021. So the later version has a lower version number - odd! Perhaps I'll try the new version against Ubuntu 20 at some point, but it's a big job to reverse out the upgrade if it doesn't work, so I'm in no hurry.

Mike

**Comment by Jean-Gabriel1982** [comment]
@oleg68 it seems to work. Do you think my parameters are correct ? (I was afraid to inverse right and left). 

I started, rebooted some times my computer and everything seems to be ok. Each time a had sound and touchscreen !!
Wahou i don't know how to thank you !! ☺️

Submenus are not tactile but I think I'll take mouse ...

**Reply by oleg68** [reply]
Congratulations. I glad to help you.
